bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java

```diff
@@ -173,47 +180,177 @@ State parseState(@Nullable String stateString) {
+                                state.getClass().getSimpleName());
+                        return false;
+                    }
+                    lhs = state instanceof Enum ? state : state.toString();
+                }
+
+                if (parsedValue == null) {
+                    // don't parse bare strings as StringType, because they are identifiers,
+                    // e.g. referring to other items
+                    List<Class<? extends State>> acceptedValueTypes = item.getAcceptedDataTypes().stream()
+                            .filter(not(StringType.class::isAssignableFrom)).toList();
+                    parsedValue = TypeParser.parseState(acceptedValueTypes, value);
+                    // Don't convert QuantityType to other types
+                    if (parsedValue != null && !(parsedValue instanceof QuantityType)) {
+                        // Try to convert it to the same type as the state
+                        // This allows comparing compatible types, e.g. PercentType vs OnOffType
+                        parsedValue = parsedValue.as(state.getClass());
+                    }
+
+                    // If the values can't be converted to a type, check to see if it's an Item name
+                    if (parsedValue == null) {
+                        try {
+                            Item valueItem = itemRegistry.getItem(value);
+                            if (valueItem != null) { // ItemRegistry.getItem can return null in tests
+                                parsedValue = valueItem.getState();
+                                // Don't convert QuantityType to other types
+                                if (!(parsedValue instanceof QuantityType)) {
+                                    parsedValue = parsedValue.as(state.getClass());
+                                }
+                                logger.debug("Condition value: '{}' is an item state: '{}' ({})", value, parsedValue,
+                                        parsedValue == null ? "null" : parsedValue.getClass().getSimpleName());
+                            }
+                        } catch (ItemNotFoundException ignore) {
+                        }
+                    }
+
+                    if (parsedValue == null) {
+                        if (comparisonType == ComparisonType.NEQ) {
+                            // They're not even type compatible, so return true for NEQ comparison
+                            return true;
+                        } else {
+                            logger.debug("Condition value: '{}' is not compatible with state '{}' ({})", value, state,
+                                    state.getClass().getSimpleName());
+                            return false;
+                        }
+                    }
+                }
+
+                rhs = Objects.requireNonNull(parsedValue instanceof StringType ? parsedValue.toString() : parsedValue);
```

Comment:
Why do you need `Objects.requireNonNull` here? If `parsedValue==null` you exit in the if-clause above.

Reply:
The compiler complained about null mismatch further down in the calls to `compareTo(rhs)`.